README.md
[![Build Status](https://travis-ci.org/sertansenturk/seyiranalyzer.svg?branch=master)](https://travis-ci.org/sertansenturk/seyiranalyzer) [![Code Climate](https://codeclimate.com/github/sertansenturk/seyiranalyzer/badges/gpa.svg)](https://codeclimate.com/github/sertansenturk/seyiranalyzer) [![GitHub version](https://badge.fury.io/gh/sertansenturk%2Fseyiranalyzer.svg)](https://badge.fury.io/gh/sertansenturk%2Fseyiranalyzer) [![DOI](https://zenodo.org/badge/21104/sertansenturk/seyiranalyzer.svg)](https://zenodo.org/badge/latestdoi/21104/sertansenturk/seyiranalyzer)

# seyir-analyzer

Tools to analyse the seyir (melodic progression) of scores and audio recordings of makam music

Introduction
=======
The repository implements the melody progression analysis methods explained in the paper.

_B. Bozkurt, "Computational analysis of overall melodic progression for Turkish Makam Music", in "Penser l’improvisation" edited by Mondher Ayari, pp. 289-298, ISBN: 9782752102485, 2015, Delatour France, Sampzon._

Currently audio seyir analysis is implemented. The algorithm gets the predominant melody of an audio recording as the input and divide it to chunks with the specified frame size and hop ratio. The pitch distribution of each chunk is computed and the peaks of these distributions are selected as the stable pitches performed at that interval. The mean pitch of each chunk is also computed. To visualize the analysis we decided to use the representation for the symbolic analysis (Figure 4), as we find it more informative.

Later score analysis and joint audio-score analysis (from alignment) will be added.

Usage
=======
```python
from seyiranalyzer.audioseyiranalyzer import AudioSeyirAnalyzer

audioSeyirAnalyzer = AudioSeyirAnalyzer()
seyir_features = audioSeyirAnalyzer.analyze(pitch, frame_dur=frame_dur, 
                                            hop_ratio=hop_ratio)

audioSeyirAnalyzer.plot(seyir_features)
```

The inputs for audioSeyirAnalyzer.analyze are:
```python
# pitch 		    : an n-by-2 numpy array, where the values in the first column are 
#					  the timestamps and the values in the second column are frequency 
#					  values. You can use the predominant melody extraction algorithm
#                     available in https://github.com/sertansenturk/predominantmelodymakam
# frame_size        : the duration of the chunks computed from the pitch in seconds
# hop_ratio         : the ratio of the time between the consecutive jumps of chunks
#                     with respect to the duration of the chunks.
#                     (Example: if the chunk_size is 16 seconds and the hop_ratio is
#                     0.25, the hop size is 4 seconds.)
```

Installation
============

If you want to install the repository, it is recommended to install the package and dependencies into a virtualenv. In the terminal, do the following:

    virtualenv env
    source env/bin/activate
    python setup.py install

If you want to be able to edit files and have the changes be reflected, then install the repository like this instead

    pip install -e .

The algorithm uses several modules in Essentia. Follow the [instructions](essentia.upf.edu/documentation/installing.html) to install the library.

Now you can install the rest of the dependencies:

    pip install -r requirements

Authors
-------
Sertan Şentürk
contact@sertansenturk.com

Reference
-------
Thesis


Changelog.md
#### seyiranalyzer v1.2.0
 - Starting the first frame "centered" around zero seconds
 - Time-stamps are generated according to the center rather than the start of the frames in plotting

#### seyiranalyzer v1.1.1
 - Updated the PitchDistribution requirement (morty)

#### seyiranalyzer v1.1.0
 - Improvements in the plot method
 - Minor refactorings

#### seyiranalyzer v1.0
 - First public release
 - Audio seyir analysis implementation


seyiranalyzer/seyiranalyzer.py
import numpy as np
import PitchDistribution
import matplotlib.pyplot as plt

class AudioSeyirAnalyzer():
    _version = "0.1"
    _slug = "makamseyir"

    def __init__(self, frame_size = 20.0, hop_ratio = 0.5, kernel_width = 7.5, 
        step_size=7.5):
        self.frame_size = frame_size  # seconds
        self.hop_size = hop_ratio * frame_size  # seconds
        self.kernel_width = kernel_width
        self.step_size = step_size

        self.dummy_ref_freq = 440.0  # hz

    def getSettings(self):
      citation = u"Bozkurt"
      
      return {'frameSize':self.frame_size, 'hopSize': self.hop_size, 
                'kernel_width': self.kernel_width, 'step_size': self.step_size,
                'citation': citation}

    def analyze(self, pitch): 
        pitch = np.array(pitch)
        tt = pitch[:,0]
        pp = pitch[:,1]

        t_st = 0
        seyir_features = []
        while t_st < tt[-1]:
            p_sliced = [p for t, p in zip(tt, pp) 
                if t_st+self.frame_size > t >= t_st]

            t_interval = [t_st, min([t_st+self.frame_size, tt[-1]])]

            p_cent = PitchDistribution.hz_to_cent(p_sliced, self.dummy_ref_freq)
            pd = PitchDistribution.generate_pd(p_cent, ref_freq=self.dummy_ref_freq,
                kernel_width=self.kernel_width, step_size=self.step_size)

            # reconvert to Hz
            pd.bins = PitchDistribution.cent_to_hz(pd.bins, pd.ref_freq)
            pd.ref_freq = []

            # normalize to 1 (instead of the area under the curve)
            maxval = max(pd.vals)
            numRatio = float(len(p_cent))/len(p_sliced) # ratio of number of samples
            timeRatio = (t_interval[1]-t_interval[0])/self.frame_size
            pd.vals = pd.vals*numRatio*timeRatio/maxval

            # get the stable pitches, i.e. peaks
            peak_idx, peak_vals = pd.detect_peaks()
            stable_pitches = [{'frequency':pd.bins[idx], 'value':val} 
                for idx, val in zip(peak_idx, peak_vals)]

            # get the average pitch
            avpitch = PitchDistribution.cent_to_hz(np.mean(p_cent), self.dummy_ref_freq)

            seyir_features.append({'pitch_distribution':pd,'stable_pitches':stable_pitches,
                'average_pitch':avpitch, 'time_interval':t_interval})

            t_st += self.hop_size

        return seyir_features

    @staticmethod
    def plot(seyir_features, plot_average_pitch=True, plot_stable_pitches=True,
             plot_distribution=False):

        if plot_distribution:
            time_starts = [sf['time_interval'][0] for sf in seyir_features]
            min_time = min(np.diff(time_starts))
            for sf in seyir_features:
                # plot the distributions through time
                yy = sf['pitch_distribution'].bins
                tt = sf['time_interval'][0] + sf['pitch_distribution'].vals * min_time*2
                plt.plot(tt, yy)

        for sf in seyir_features:
            t_st = sf['time_interval'][0]
            max_peak = max([sp['value'] for sp in sf['stable_pitches']])
            for sp in sf['stable_pitches']:
                clr = 'r' if sp['value'] == max_peak else 'b'
                # map the values from 0-1 to 1-6
                marker_thickness = sp['value']*5+1 
                plt.plot(t_st, sp['frequency'], 'o', color = clr, ms=marker_thickness)

        if plot_average_pitch:
            tt = [sf['time_interval'][0] for sf in seyir_features]
            pp = [sf['average_pitch'] for sf in seyir_features]

            plt.plot(tt, pp, color='k', linewidth=3)

        plt.xlim([seyir_features[0]['time_interval'][0], 
            seyir_features[-1]['time_interval'][1]])

        plt.show()

seyiranalyzer/PitchDistribution.py
# -*- coding: utf-8 -*-
import essentia
import essentia.standard as std
import numpy as np
import json
import os
from scipy.integrate import simps
from scipy.stats import norm
import matplotlib.pyplot as plt

def load(fname):
	"""-------------------------------------------------------------------------
	Loads a PitchDistribution object from JSON file.
	----------------------------------------------------------------------------
	fname    : The filename of the JSON file
	-------------------------------------------------------------------------"""
	with open(fname) as f:
		dist = json.load(f)

	return PitchDistribution(np.array(dist[0]['bins']), np.array(dist[0]['vals']),
		                     kernel_width=dist[0]['kernel_width'], ref_freq=dist[0]['ref_freq'])

class PitchDistribution:
	
	def __init__(self, pd_bins, pd_vals, kernel_width=7.5, ref_freq=440):
		"""------------------------------------------------------------------------
		The main data structure that wraps all the relevant information about a 
		pitch distribution.
		---------------------------------------------------------------------------
		pd_bins      : Bins of the pitch distribution. It is a 1-D list of equally
		               spaced monotonically increasing frequency values.
		pd_vals      : Values of the pitch distribution
		kernel_width : The std. deviation of the Gaussian kernel. See generate_pd()
		               of ModeFunctions for more detail.
		ref_freq     : Reference frequency that is used while generating the 
		               distribution. If the tonic of a recording is annotated,
		               this is variable that stores it.
		-------------------------------------------------------------------------"""
		self.bins = pd_bins
		self.vals = pd_vals
		self.kernel_width = kernel_width
		self.ref_freq = ref_freq

		### Due to the floating point issues in Python, the step_size might not be
		### exactly equal to (for example) 7.5, but 7.4999... In such cases the 
		### bin generation of pitch distributions include 1200 cents too and chaos
		### reigns. We fix it here.
		temp_ss = self.bins[1] - self.bins[0]
		self.step_size = (temp_ss if temp_ss == (round(temp_ss * 10) / 10) 
			else round(temp_ss * 10) / 10)

	def save(self, fname, save_dir='./'):
		"""-------------------------------------------------------------------------
		Saves the PitchDistribution object to a JSON file.
		----------------------------------------------------------------------------
		fname    : The name of the JSON file to be created.
		save_dir : Pathway of where the JSON would saved
		-------------------------------------------------------------------------"""
		dist_json = [{'bins':self.bins.tolist(), 'vals':self.vals.tolist(),
		              'kernel_width':self.kernel_width, 'ref_freq':self.ref_freq, 
		              'step_size':self.step_size}]
		with open(os.path.join(save_dir, fname), 'w') as f:
			json.dump(dist_json, f, indent=2)
			f.close()

	def is_pcd(self):
		"""-------------------------------------------------------------------------
		The boolean flag of whethwe the instance is PCD or not.
		-------------------------------------------------------------------------"""
		return (max(self.bins) == (1200 - self.step_size) and min(self.bins) == 0)

	def detect_peaks(self):
		"""-------------------------------------------------------------------------
		Finds the peak indices of the distribution. These are treated as tonic
		candidates in higher order functions.
		-------------------------------------------------------------------------"""
		# Peak detection is handled by Essentia
		detector = std.PeakDetection()
		peak_bins, peak_vals = detector(essentia.array(self.vals))

		# Essentia normalizes the positions to 1, they are converted here
		# to actual index values to be used in bins.
		peak_idxs = [round(bn * (len(self.bins) - 1)) for bn in peak_bins]
		if(peak_idxs[0] == 0):
			peak_idxs = np.delete(peak_idxs, [len(peak_idxs) - 1])
		
		peak_vals = [self.vals[idx] for idx in peak_idxs]
		return peak_idxs, peak_vals

	def shift(self, shift_idx):
		"""-------------------------------------------------------------------------
		Shifts the distribution by the given number of samples
		----------------------------------------------------------------------------
		shift_idx : The number of samples that the distribution is tı be shifted
		-------------------------------------------------------------------------"""
		# If the shift index is non-zero, do the shifting procedure
		if(shift_idx):
			
			# If distribution is a PCD, we do a circular shift
			if self.is_pcd():
				shifted_vals = np.concatenate((self.vals[shift_idx:], self.vals[:shift_idx]))
			
			# If distribution is a PD, it just shifts the values. In this case,
			# pd_zero_pad() of ModeFunctions is always applied beforehand to make
			# sure that no non-zero values are dropped.
			else:
				
				# Shift towards left
				if(shift_idx > 0):
					shifted_vals = np.concatenate((self.vals[shift_idx:], np.zeros(shift_idx)))
				
				# Shift towards right
				else: 
					shifted_vals = np.concatenate((np.zeros(abs(shift_idx)), self.vals[:shift_idx]))

			return PitchDistribution(self.bins, shifted_vals, kernel_width=self.kernel_width,
				ref_freq=self.ref_freq)
		
		# If a zero sample shift is requested, a copy of the original distribution
		# is returned
		else:
			return PitchDistribution(self.bins, self.vals, kernel_width=self.kernel_width,
				ref_freq=self.ref_freq)

	def plot(self):
		plt.plot(self.bins, self.vals)

def generate_pd(cent_track, ref_freq=440, kernel_width=7.5, step_size=7.5):
	"""-------------------------------------------------------------------------
	Given the pitch track in the unit of cents, generates the Pitch Distribution
	of it. the pitch track from a text file. 0th column is the time-stamps and
	1st column is the corresponding frequency values.
	----------------------------------------------------------------------------
	cent_track:     1-D array of frequency values in cents.
	ref_freq:       Reference frequency used while converting Hz values to cents.
	                This number isn't used in the computations, but is to be
	                recorded in the PitchDistribution object.
	kernel_width:  The standard deviation of the gaussian kernel, used in Kernel
	                Density Estimation. If 0, a histogram is given
	step_size:        The step size of the Pitch Distribution bins.
	-------------------------------------------------------------------------"""

	### Some extra interval is added to the beginning and end since the
	### superposed Gaussian for kernel_width would introduce some tails in the
	### ends. These vanish after 3 sigmas(=kernel_width).

	### The limits are also quantized to be a multiple of chosen step-size
	### kernel_width = standard deviation fo the gaussian kernel

	### TODO: filter out the NaN, -infinity and +infinity from the pitch track

	# Finds the endpoints of the histogram edges. Histogram bins will be
	# generated as the midpoints of these edges. 
	min_edge = min(cent_track) - (step_size / 2.0) - kernel_width*3
	max_edge = max(cent_track) + (step_size / 2.0) + kernel_width*3
	pd_edges = np.concatenate([np.arange(-step_size/2.0, min_edge, -step_size)[::-1],
	                           np.arange(step_size/2.0, max_edge, step_size)])

	# An exceptional case is when min_bin and max_bin are both positive or negative
	# In this case, remove unnecessary bins

	if min_edge > 0:
		pd_edges = [e for e in pd_edges if e >= min_edge]

	if max_edge < 0:
		pd_edges = [e for e in pd_edges if e <= max_edge]
	
	# Generates the histogram and bins (i.e. the midpoints of edges)
	pd_vals, pd_edges = np.histogram(cent_track, bins=pd_edges, density=True)
	pd_bins = np.convolve(pd_edges, [0.5,0.5])[1:-1]

	if kernel_width > 0: # kernel density estimation (approximated)
		# smooth the histogram
		normal_dist = norm(loc = 0, scale = kernel_width)
		xn = np.concatenate([np.arange(0, - 5 * kernel_width, -step_size)[::-1], 
		    np.arange(step_size, 5 * kernel_width, step_size)])
		sampled_norm = normal_dist.pdf(xn)

		extra_num_bins = len(sampled_norm)/2 # convolution generates tails
		pd_vals = np.convolve(pd_vals, sampled_norm)[extra_num_bins:-extra_num_bins]

		# normalize the area under the curve
		area = simps(pd_vals, dx=step_size)
		pd_vals = pd_vals/area

	# Sanity check. If the histogram bins and vals lengths are different, we
	# are in trouble. This is an important assumption of higher level functions.
	if(len(pd_bins) != len(pd_vals)):
		raise ValueError('Lengths of bins and Vals are different')

	# Initializes the PitchDistribution object and returns it.
	return PitchDistribution(pd_bins, pd_vals, kernel_width=kernel_width, ref_freq=ref_freq)

def generate_pcd(pd):
	"""-------------------------------------------------------------------------
	Given the pitch distribution of a recording, generates its pitch class
	distribution, by octave wrapping.
	----------------------------------------------------------------------------
	pd: PitchDistribution object. Its attributes include everything we need
	-------------------------------------------------------------------------"""

	# Initializations
	pcd_bins = np.arange(0, 1200, pd.step_size)
	pcd_vals = np.zeros(len(pcd_bins))

	# Octave wrapping
	for k in range(len(pd.bins)):
		idx = int((pd.bins[k] % 1200) / pd.step_size)
		idx = idx if idx != 160 else 0
		pcd_vals[idx] += pd.vals[k]

	# Initializes the PitchDistribution object and returns it.
	return (pcd_bins, pcd_val)


def hz_to_cent(hz_track, ref_freq):
	"""-------------------------------------------------------------------------
	Converts an array of Hertz values into cents.
	----------------------------------------------------------------------------
	hz_track : The 1-D array of Hertz values
	ref_freq	: Reference frequency for cent conversion
	-------------------------------------------------------------------------"""
	hz_track = np.array(hz_track)

	# The 0 Hz values are removed, not only because they are meaningless,
	# but also logarithm of 0 is problematic.
	return np.log2(hz_track[hz_track>0] / ref_freq) * 1200.0


def cent_to_hz(cent_track, ref_freq):
	"""-------------------------------------------------------------------------
	Converts an array of cent values into Hertz.
	----------------------------------------------------------------------------
	cent_track  : The 1-D array of cent values
	ref_freq	: Reference frequency for cent conversion
	-------------------------------------------------------------------------"""
	cent_track = np.array(cent_track)

	return 2 ** (cent_track / 1200.0) * ref_freq


seyiranalyzer/AudioSeyirAnalyzer.py
# -*- coding: utf-8 -*-
import numpy as np
import matplotlib.pyplot as plt
from modetonicestimation.PitchDistribution import PitchDistribution
from modetonicestimation.Converter import Converter


class AudioSeyirAnalyzer(object):
    _dummy_ref_freq = 440.0  # hz
    citation = u"B. Bozkurt, Computational analysis of overall melodic " \
               u"progression for Turkish Makam Music, in Penser " \
               u"l’improvisation edited by Mondher Ayari, pp. 289-298, " \
               u"ISBN: 9782752102485, 2015, Delatour France, Sampzon."

    def __init__(self, kernel_width=7.5, step_size=7.5):
        self.kernel_width = kernel_width
        self.step_size = step_size

    def _get_settings(self):
        return {'kernel_width': self.kernel_width, 'step_size': self.step_size,
                'citation': self.citation}

    def analyze(self, pitch, frame_dur=20.0, hop_ratio=0.5):
        hop_size = frame_dur * hop_ratio

        pitch = np.array(pitch)
        tt = pitch[:, 0]
        pp = pitch[:, 1]

        tb = 0
        t_intervals = []
        while tb < tt[-1]:
            t_intervals.append([tb, min([tb + frame_dur, tt[-1]])])
            tb += hop_size

        return self._compute_seyir_features_per_interval(pp, tt, t_intervals)

    def _compute_seyir_features_per_interval(self, pp, tt, t_intervals):
        seyir_features = []
        maxdur = 0
        for ti in t_intervals:
            dur = ti[1] - ti[0]
            if dur > maxdur:
                maxdur = dur

        for ti in t_intervals:
            p_sliced = [p for t, p in zip(tt, pp) if ti[1] > t >= ti[0]]
            p_cent = Converter.hz_to_cent(p_sliced, self._dummy_ref_freq,
                                          min_freq=20.0)

            # pop nan and inf
            p_cent = p_cent[~np.isnan(p_cent)]
            p_cent = p_cent[~np.isinf(p_cent)]  # shouldnt exist but anyways...

            if p_cent.size == 0:  # silence
                seyir_features.append(
                    {'pitch_distribution': [], 'stable_pitches': [],
                     'average_pitch': np.nan, 'time_interval': ti})
            else:
                pd = PitchDistribution.from_cent_pitch(
                    p_cent, ref_freq=self._dummy_ref_freq,
                    smooth_factor=self.kernel_width, step_size=self.step_size)

                # reconvert to Hz
                pd.cent_to_hz()

                # normalize to 1 (instead of the area under the curve)
                maxval = max(pd.vals)
                num_ratio = float(len(p_cent)) / len(p_sliced)  # ratio of
                # number of samples
                time_ratio = (ti[1] - ti[0]) / maxdur
                pd.vals = pd.vals * num_ratio * time_ratio / maxval

                # get the stable pitches, i.e. peaks
                peak_idx, peak_vals = pd.detect_peaks()
                stable_pitches = [{'frequency': float(pd.bins[idx]),
                                   'value': float(val)}
                                  for idx, val in zip(peak_idx, peak_vals)]

                # get the average pitch
                avpitch = Converter.cent_to_hz(np.mean(p_cent),
                                               self._dummy_ref_freq)

                seyir_features.append(
                    {'pitch_distribution': pd,
                     'stable_pitches': stable_pitches,
                     'average_pitch': avpitch, 'time_interval': ti})

        return seyir_features

    @staticmethod
    def plot(seyir_features, ax=None, plot_average_pitch=True,
             plot_stable_pitches=True, plot_distribution=False):

        if ax is None:
            fig, ax = plt.subplots(1, 1)

        if plot_distribution:
            time_starts = [sf['time_interval'][0] for sf in seyir_features]
            min_time = min(np.diff(time_starts))
            for sf in seyir_features:
                if sf['pitch_distribution']:  # ignore silent frame
                    # plot the distributions through time
                    yy = sf['pitch_distribution'].bins
                    tt = (sf['time_interval'][0] +
                          sf['pitch_distribution'].vals * min_time * 2)
                    ax.plot(tt, yy)

        if plot_stable_pitches:
            num_frames = len(seyir_features)
            for sf in seyir_features:
                if sf['stable_pitches']:  # ignore silent frame
                    t_st = sf['time_interval'][0]
                    max_peak = max([sp['value']
                                    for sp in sf['stable_pitches']])
                    for sp in sf['stable_pitches']:
                        clr = 'r' if sp['value'] == max_peak else 'b'
                        # map the values from 0-1 to 1-6
                        marker_thickness = ((sp['value'] * 5 + 1) * 100 /
                                            num_frames)
                        ax.plot(t_st, sp['frequency'], 'o',
                                color=clr, ms=marker_thickness)

        if plot_average_pitch:
            tt = [sf['time_interval'][0] for sf in seyir_features]
            pp = [sf['average_pitch'] for sf in seyir_features]

            ax.plot(tt, pp, color='k', linewidth=3)

        ax.set_xlim([seyir_features[0]['time_interval'][0],
                     seyir_features[-1]['time_interval'][1]])
        ax.set_xlabel('Time (sec)')
        ax.set_ylabel('Frequency (Hz)')


seyiranalyzer/audioseyiranalyzer.py
# -*- coding: utf-8 -*-

# Copyright 2016 Sertan Şentürk
#
# This file is part of seyiranalyzer
#
# seyiranalyzer is free software: you can redistribute it and/or modify it
# under the terms of the GNU Affero General Public License as published by
# the Free Software Foundation (FSF), either version 3 of the License,
# or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
# details.
#
# You should have received a copy of the GNU General Public License along with
# this program.  If not, see http://www.gnu.org/licenses/
import numpy as np
import matplotlib.pyplot as plt
from morty.pitchdistribution import PitchDistribution
from morty.converter import Converter
import json
import copy


class AudioSeyirAnalyzer(object):
    _dummy_ref_freq = 440.0  # hz
    citation = u"B. Bozkurt, Computational analysis of overall melodic " \
               u"progression for Turkish Makam Music, in Penser " \
               u"l’improvisation edited by Mondher Ayari, pp. 289-298, " \
               u"ISBN: 9782752102485, 2015, Delatour France, Sampzon."

    def __init__(self, kernel_width=7.5, step_size=7.5):
        self.kernel_width = kernel_width
        self.step_size = step_size

    def _get_settings(self):
        return {'kernel_width': self.kernel_width, 'step_size': self.step_size,
                'citation': self.citation}

    def analyze(self, pitch, frame_dur=20.0, hop_ratio=0.5):
        hop_size = frame_dur * hop_ratio

        pitch = np.array(pitch)
        tt = pitch[:, 0]
        pp = pitch[:, 1]

        # start the first frame "centered" around 0 seconds
        tb = -frame_dur / 2.0
        t_intervals = []
        t_center = []
        while tb < tt[-1]:
            # make sure the frame boundaries do not exceed the audio time
            t_intervals.append([max([0.0, tb]), min([tb + frame_dur, tt[-1]])])
            t_center.append(min([max([tb + frame_dur / 2.0, 0]), tt[-1]]))
            tb += hop_size

        return self._compute_seyir_features_per_interval(pp, tt, t_intervals,
                                                         t_center)

    def _compute_seyir_features_per_interval(self, pp, tt, t_intervals,
                                             t_center):
        seyir_features = []
        maxdur = max(ti[1] - ti[0] for ti in t_intervals)

        for ti, tc in zip(t_intervals, t_center):
            p_cent, p_sliced = self._slice_pitch(pp, ti, tt)

            if p_cent.size == 0:  # silence
                seyir_features.append(
                    {'pitch_distribution': [], 'average_pitch': np.nan,
                     'stable_pitches': [], 'time_interval': ti,
                     'time_center': tc})
            else:
                pd = PitchDistribution.from_cent_pitch(
                    p_cent, ref_freq=self._dummy_ref_freq,
                    kernel_width=self.kernel_width, step_size=self.step_size)

                # reconvert to Hz
                pd.cent_to_hz()

                # normalize to 1 (instead of the area under the curve)
                maxval = max(pd.vals)
                num_ratio = float(len(p_cent)) / len(p_sliced)  # ratio of
                # number of samples
                time_ratio = (ti[1] - ti[0]) / maxdur
                pd.vals = pd.vals * num_ratio * time_ratio / maxval

                # get the stable pitches, i.e. peaks
                peak_idx, peak_vals = pd.detect_peaks()
                stable_pitches = [{'frequency': float(pd.bins[idx]),
                                   'value': float(val)}
                                  for idx, val in zip(peak_idx, peak_vals)]

                # get the average pitch
                avpitch = Converter.cent_to_hz(np.mean(p_cent),
                                               self._dummy_ref_freq)

                seyir_features.append(
                    {'pitch_distribution': pd, 'average_pitch': avpitch,
                     'stable_pitches': stable_pitches, 'time_interval': ti,
                     'time_center': tc})

        return seyir_features

    def _slice_pitch(self, pp, ti, tt):
        p_sliced = [p for t, p in zip(tt, pp) if ti[1] > t >= ti[0]]
        p_cent = Converter.hz_to_cent(p_sliced, self._dummy_ref_freq,
                                      min_freq=20.0)
        # pop nan and inf
        p_cent = p_cent[~np.isnan(p_cent)]
        p_cent = p_cent[~np.isinf(p_cent)]  # shouldn't exist, but anyways...
        return p_cent, p_sliced

    @staticmethod
    def to_json(seyir_features, json_out=None):
        seyir_copy = copy.deepcopy(seyir_features)
        AudioSeyirAnalyzer.serialize(seyir_copy)

        if json_out is None:
            return json.dumps(seyir_copy)
        else:
            json.dump(seyir_copy, open(json_out, 'w'))

    @staticmethod
    def from_json(json_in):
        try:
            seyir_features = json.load(open(json_in, 'r'))
        except IOError:
            seyir_features = json.loads(json_in)

        AudioSeyirAnalyzer.deserialize(seyir_features)

        return seyir_features

    @staticmethod
    def serialize(seyir_features):
        for sf in seyir_features:
            try:  # convert pitch distribution objects to dicts
                sf['pitch_distribution'] = sf[
                    'pitch_distribution'].to_dict()
            except AttributeError:  # empty pitch distribution
                assert not sf['pitch_distribution'], \
                    'non-empty, non-object pitch distribution encountered'

    @staticmethod
    def deserialize(seyir_features):
        for sf in seyir_features:
            try:
                sf['pitch_distribution'] = PitchDistribution.from_dict(
                    sf['pitch_distribution'])
            except AttributeError:  # empty pitch distribution
                assert not sf['pitch_distribution'], \
                    'non-empty, non-object pitch distribution encountered'

    @staticmethod
    def plot(seyir_features, ax=None, plot_average_pitch=True,
             plot_stable_pitches=True, plot_distribution=False):

        if ax is None:
            fig, ax = plt.subplots(1, 1)

        if plot_distribution:
            AudioSeyirAnalyzer._pitch_distrib_plotter(ax, seyir_features)

        if plot_stable_pitches:
            AudioSeyirAnalyzer._stable_pitch_plotter(ax, seyir_features)

        if plot_average_pitch:
            AudioSeyirAnalyzer._average_pitch_plotter(ax, seyir_features)

        ax.set_xlim([seyir_features[0]['time_interval'][0],
                     seyir_features[-1]['time_interval'][1]])
        ax.set_xlabel('Time (sec)')
        ax.set_ylabel('Frequency (Hz)')

    @staticmethod
    def _average_pitch_plotter(ax, seyir_features):
        tt = [sf['time_center'] for sf in seyir_features]
        pp = [sf['average_pitch'] for sf in seyir_features]
        ax.plot(tt, pp, color='k', linewidth=3)

    @staticmethod
    def _stable_pitch_plotter(ax, seyir_features):
        num_frames = len(seyir_features)
        for sf in seyir_features:
            if sf['stable_pitches']:  # ignore silent frame
                t_center = sf['time_center']
                max_peak = max([sp['value']
                                for sp in sf['stable_pitches']])
                for sp in sf['stable_pitches']:
                    clr = 'r' if sp['value'] == max_peak else 'b'
                    # map the values from 0-1 to 1-6
                    marker_thickness = ((sp['value'] * 5 + 1) * 100 /
                                        num_frames)
                    ax.plot(t_center, sp['frequency'], 'o',
                            color=clr, ms=marker_thickness)

    @staticmethod
    def _pitch_distrib_plotter(ax, seyir_features):
        time_centers = [sf['time_center'][0] for sf in seyir_features]
        min_time = min(np.diff(time_centers))
        for sf in seyir_features:
            if sf['pitch_distribution']:  # ignore silent frame
                # plot the distributions through time
                yy = sf['pitch_distribution'].bins
                tt = (sf['time_center'] +
                      sf['pitch_distribution'].vals * min_time * 2)
                ax.plot(tt, yy)
